client/lclusterc.go

```diff
@@ -0,0 +1,324 @@
+package main
+
+import (
+    "fmt"
+    "flag"
+    "time"
+    "strconv"
+    "golang.org/x/net/context"
+    grpc "google.golang.org/grpc"
+    pb "../lclusterpb"
+)
+
+// Variables to hold the result of argument flags.
+var addjob string
+var checkjob string
+var removejob string
+
+// Initialize the flags beforehand.
+func init() {
+
+    // Argument flag for when end-user requests to add a job
+    flag.StringVar(&addjob, "addjob", "",
+      "Commandline program to execute; e.g. 'grep abc /path/to/file' ")
+
+    // Argument flag for when the end-user wants to check on a job
+    flag.StringVar(&checkjob, "checkjob", "",
+      "Uuid of the job to query status.")
+
+    // Argument flag for when the end-user wants to remove a job
+    flag.StringVar(&removejob, "removejob", "",
+      "Pid of the job to be removed.")
+}
+
+//! Add a job to the server.
+/*
+ * @param    string    commandline to execute
+ *
+ * @return   none
+ */
+func addJobToServer(cmd string) {
+
+    // Input validation
+    if len(cmd) < 1 {
+        fmt.Printf("addJobToServer() --> invalid input")
+        return
+    }
+
+    // Dial a connection to the grpc server.
+    connection, err := grpc.Dial(grpcServerAddr + grpcPort,
+      grpc.WithInsecure())
+
+    // Safety check, ensure no errors have occurred.
+    if err != nil {
+        fmt.Printf("addJobToServer() --> unable to connect to grpc " +
+                   "server at address")
+        return
+    }
+
+    // Defer the connection for the time being, but eventually it will be
+    // closed once we've finished with it.
+    defer connection.Close()
+
+    // Create an lcluster client
+    lcluster_client := pb.NewLclusterdClient(connection)
+
+    // Start a new job request using the data obtained above.
+    request := &pb.StartJobRequest{
+		Path:     cmd,
+		Args:     []string{sh, cmd},
+		Env:      []string{"PATH=/bin"},
+		Hostname: grpcServerAddr,
+    }
+
+    // Grab the current background context.
+    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+
+    // Using the new job request defined above, go ahead and 
+    response, err := lcluster_client.StartJob(ctx, request)
+
+    // Cancel the current context since this has either generated a
+    // response or an error.
+    cancel()
+
+    // Safety check, ensure that no error has occurred.
+    if err != nil || response.GetPid() < 0 {
+        fmt.Printf("\nError: Unable to add job to queue!\n")
+        fmt.Printf(err.Error() + "\n")
+        return
+    }
+
+    // Since the job was started properly, go ahead and print out a message
+    // telling the end-user the uuid of the new job.
+    fmt.Printf("The requested job has been added to queue, has pid of " +
+      strconv.FormatInt(response.GetPid(), 10) + "\n")
+}
+
+//! Status of a job from the server.
+/*
+ * @param    string    uuid of job to stat
+ *
+ * @return   none
+ */
+func checkJobOnServer(uuid string) {
+
+    // Input validation
+    if len(uuid) < 1 {
```

Comment:
Ideally, I think you make a good case for having the errors pass back as much as possible, rather than simply getting dumped into stdout. I have since resolved most of the worst cases, but I have decided to let some of the node manager function still dump to stdout since I still want the server to give the terminal some meaningful output.

https://github.com/rbisewski/lclusterd/commit/4c084e2f770f10f30be3241cfa33e0fac9968f67